File: Ruby/game.rb
require_relative 'board'
require_relative 'human_player'
require_relative 'computer_player'

class Game
    attr_reader :board
    def initialize
        @p1 = HumanPlayer.new("Human", :X)
        @p2 = ComputerPlayer.new("Computer", :O)
        @current_player = @p1
        @board = Board.new
    end

    def current_player
        @current_player
    end

    def switch_players!
        if @current_player == @p1
            @current_player = @p2
        else
            @current_player = @p1
        end
    end

    def play_turn
        puts @board
        move = @current_player.get_move(@board)
        mark = @current_player.mark
        @board.place_mark(move, mark)
        switch_players!
    end

    def start
        puts "______________________________________"
        puts "Let's play Tic Tac Toe with Minimax AI"
        puts "______________________________________"
        until @board.over?
            play_turn
        end

        if @board.winner.nil?
            puts "Game over, it's a tied"
        else
            puts "Game over, and winner is #{@board.winner}"
        end
    end
end

game = Game.new
game.start

File: Golang/src/game.go
package main

import "fmt"

func NewGame() Game {
	board := NewBoard()
	p1 := NewPlayer("Human", board)
	p2 := NewPlayer("Computer", board)
	return Game{p1, p2, p1, board}
}

type Game struct {
	PlayerOne Player
	PlayerTwo Player
	CurrentPlayer Player
	Board Board
}

func (game *Game) Start() {
	for !game.Board.IsOver() {
		i, j := game.CurrentPlayer.GetMove()

		if (game.CurrentPlayer == game.PlayerOne) {
			game.Board.PlaceMark(i, j, "X")
		} else {
			game.Board.PlaceMark(i, j, "O")
		}

		fmt.Println(game.Board)
		game.switchPlayer()
	}
}